04.ls/ls.rb

```diff
@@ -33,4 +33,118 @@ def split_files(files)
   divided_files
 end
 
+FILE_TYPE = {
+  'fifo' => 'p',
+  'characterSpecial' => 'c',
+  'directory' => 'd',
+  'blockSpecial' => 'b',
+  'file' => '-',
+  'link' => 'l',
+  'socket' => 's'
+}.freeze
+
+PERMISSION = {
+  '0' => '---',
+  '1' => '--x',
+  '2' => '-w-',
+  '3' => '-wx',
+  '4' => 'r--',
+  '5' => 'r-x',
+  '6' => 'rw-',
+  '7' => 'rwx'
+}.freeze
+
+def load_l_option(path, file_names)
+  stat_files = file_names.map { |file_name| File.lstat("#{path}/#{file_name}") }
+  path_to_stat = stat_files.map.with_index { |stat, idx| ["#{path}/#{file_names[idx]}", stat] }.to_h
+
+  puts "total #{stat_files.sum(&:blocks)}"
+  load_body(path_to_stat).each do |file_show_data|
+    puts format_file(file_show_data, stat_files)
+  end
+end
+
+def load_body(path_to_stat)
+  path_to_stat.map do |absolute_path, stat|
+    {
+      permission: to_permission(stat),
+      link_count: stat.nlink.to_s,
+      owner: Etc.getpwuid(stat.uid).name,
+      group: Etc.getgrgid(stat.gid).name,
+      byte_size: stat.size.to_s,
+      time_stamp: stat.mtime.strftime(to_timestamp(stat.mtime)),
+      file_name: to_file_name(stat, absolute_path)
+    }
+  end
+end
+
+def to_timestamp(mtime)
+  branch_sec = 15_552_000
+  (Time.now - mtime) >= branch_sec ? ' %_m %_d  %Y ' : ' %_m %_d %H:%M '
+end
+
+def to_file_name(stat, absolute_path)
+  file_name = File.basename(absolute_path)
+  stat.symlink? ? "#{file_name} -> #{File.readlink(absolute_path)}" : file_name
+end
+
+def format_file(file_show_data, stat_files)
+  section_to_maxlength = fetch_maxlength(stat_files)
+  space = ' '
+  [
+    file_show_data[:permission], space * 2,
+    file_show_data[:link_count].rjust(section_to_maxlength[:link]), space,
+    file_show_data[:owner].ljust(section_to_maxlength[:owner]), space * 2,
+    file_show_data[:group].ljust(section_to_maxlength[:group]), space * 2,
+    file_show_data[:byte_size].rjust(section_to_maxlength[:byte_size]),
+    file_show_data[:time_stamp],
+    file_show_data[:file_name]
+  ].join
+end
+
+def fetch_maxlength(stat_files)
+  section_to_length = { link: 0, owner: 0, group: 0, byte_size: 0 }
+
+  stat_files.each do |stat|
+    section_to_length[:link] = stat.nlink.to_s.length if section_to_length[:link] < stat.nlink.to_s.length
```

Comment:
この実装でもまだ遅いんじゃないでしょうか。
たとえば、以下のようにスリープしたとき、対象ファイルが10件あるなら、sleep時間は全体で1秒で収まるようにしてほしいです。（ヒント：修正するのはこのメソッドのロジックではありません）

```suggestion
  stat_files.each do |stat|
    sleep 0.1
    section_to_length[:link] = stat.nlink.to_s.length if section_to_length[:link] < stat.nlink.to_s.length
```

Reply:
下記のように、ロジックはそのままにそれぞれメソッド分割しました。
一気に呼び出すのではなく、「必要になった時に各自取得する」という形にしました。

```ruby
# lengthは初期値, 0が代入されています。
def maxlength_link(stat_files, length)
  stat_files.each do |stat|
    length = stat.nlink.to_s.length if length < stat.nlink.to_s.length
  end
  length
end
```

修正前のロジックでは、一箇所(例えば`link`の最大文字取得)が遅れるだけで、
`link,uid,gid,byte_size`全ての最大文字取得の処理が足を引っ張られてしまう。
だから分割しなさい、ということなのかな？と考え、メソッドの分割をするに至りました。


質問があります！
そもそもなぜここでsleepメソッドを使うのでしょうか？ということをお聞きしたいです。
* 特定の処理が終わるまで一時停止する。
* 間隔をあけて処理を実行する。

`sleep`メソッドはこのような状況で使うものという認識です。

この場で使いたくなるケースに思い至れず困惑しています。
的外れな修正であれば申し訳ないです🙇‍♂️